pay/contracts/v1/HectorPay.sol

```diff
@@ -0,0 +1,693 @@
+            to,
+            amountPerSec,
+            starts,
+            ends,
+            starts
+        );
+        emit StreamCreated(
+            msg.sender,
+            to,
+            amountPerSec,
+            starts,
+            ends,
+            streamId
+        );
+    }
+
+    function createStreamWithReason(
+        address to,
+        uint256 amountPerSec,
+        uint48 starts,
+        uint48 ends,
+        string calldata reason
+    ) public onlyActiveSubscription(msg.sender) {
+        bytes32 streamId = _createStream(
+            to,
+            amountPerSec,
+            starts,
+            ends,
+            starts
+        );
+        emit StreamCreatedWithReason(
+            msg.sender,
+            to,
+            amountPerSec,
+            starts,
+            ends,
+            streamId,
+            reason
+        );
+    }
+
+    function deposit(uint256 amount) public nonReentrant {
+        uint256 decimalAmount = amount * DECIMALS_DIVISOR;
+
+        payers[msg.sender].totalDeposited += decimalAmount;
+
+        token.safeTransferFrom(msg.sender, address(this), amount);
+
+        emit PayerDeposit(msg.sender, decimalAmount);
+    }
+
+    function depositAndCreate(
+        uint256 amountToDeposit,
+        address to,
+        uint256 amountPerSec,
+        uint48 starts,
+        uint48 ends
+    ) external {
+        deposit(amountToDeposit);
+        createStream(to, amountPerSec, starts, ends);
+    }
+
+    function depositAndCreateWithReason(
+        uint256 amountToDeposit,
+        address to,
+        uint256 amountPerSec,
+        uint48 starts,
+        uint48 ends,
+        string calldata reason
+    ) external {
+        deposit(amountToDeposit);
+        createStreamWithReason(to, amountPerSec, starts, ends, reason);
+    }
+
+    function withdraw(
+        address from,
+        address to,
+        uint256 amountPerSec,
+        uint48 starts,
+        uint48 ends
+    ) external {
+        (, , uint256 amountToTransfer) = _withdraw(
+            from,
+            to,
+            amountPerSec,
+            starts,
+            ends
+        );
+        token.safeTransfer(to, amountToTransfer);
+    }
+
+    function cancelStream(
+        address to,
+        uint256 amountPerSec,
+        uint48 starts,
+        uint48 ends
+    ) external {
+        bytes32 streamId = getStreamId(
+            msg.sender,
+            to,
+            amountPerSec,
+            starts,
+            ends
+        );
+
+        /// resume stream if it's paused in order to cancel it
+        if (
+            streams[streamId].lastPaid == 0 && streams[streamId].lastPaused > 0
+        ) {
+            _resumeStream(msg.sender, to, amountPerSec, starts, ends);
+        }
+
+        /// cancel stream
+        _cancelStream(msg.sender, to, amountPerSec, starts, ends);
+
+        /// calculate total committed amount of a stream
+        uint256 start = max(starts, block.timestamp);
```

Comment:
@bejaxer is it possible to move the "state-changing" logic from line 546-549 before line 543 which triggers the transfer call?
Reason: State variables written after the call(s) - https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities